### PR TITLE
First part of ZONEMD support

### PR DESCRIPTION
**Pull request by omoerbeek**
### Short description

1) Define record type
2) Verification according to RFC, including many failure cases and RSIG handling. Support sha384 and sha512
3) Unit tests, including all examples from RFC and a few failure cases.
4) Command line for pdnsutil: zonemd-verify-file

Based on POC from @Habbie, with many changes.

Things to be done I can think of (details to be decided):
 
For the Recursor: 
- validation of incoming zones (potentially Zone to Cache, RPZ, Auth-Zones). Those cases need to consider DNSSEC validation of ZONEMD records as well per RFC. Also config of zone should indicate if ZONEMD verification is required. For Zone to Cache and RPZ we can do that from Lua, for Auth Zones it is not immediately clear how.

For the Authoritative Server:
- verify/insert/update/remove ZONEMD in DB 
- send ZONEMD out on AXFR. 
- verify incoming AXFR ZONEMD. 

Both should be handled in separate PRs imo.

### Checklist

I have:
- [X] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [X] compiled this code
- [X] tested this code
- [X] included documentation (including possible behaviour changes)
- [X] documented the code
- [ ] added or modified regression test(s)
- [X] added or modified unit test(s)

### CODE REVIEW

File: pdns/sha.hh [hunk 1]
```diff
@@ -51,3 +52,66 @@ inline std::string pdns_sha512sum(const std::string& input)
...
+        md = EVP_sha256();
+        break;
+      case 384:
+        md = EVP_sha384();
+        break;
+      case 512:
+        md = EVP_sha512();
+        break;
+      default:
+        throw std::runtime_error("SHADigest: unsupported size");
```
Comment (pieterlexis) [inline pdns/sha.hh]: maybe use `std::out_of_range`?

File: pdns/sha.hh [hunk 2]
```diff
@@ -51,3 +52,67 @@ inline std::string pdns_sha512sum(const std::string& input)
   SHA512(reinterpret_cast<const unsigned char*>(input.c_str()), input.length(), result);
   return std::string(result, result + sizeof result);
 }
+
+namespace pdns
+{
+class SHADigest
+{
+public:
+  SHADigest(int bits)
```
Comment (rgacogne) [inline pdns/sha.hh:61]:
```suggestion
  SHADigest(unsigned int bits)
```

File: pdns/sha.hh [hunk 3]
```diff
@@ -51,3 +52,67 @@ inline std::string pdns_sha512sum(const std::string& input)
...
+      md = EVP_sha256();
+      break;
+    case 384:
+      md = EVP_sha384();
+      break;
+    case 512:
+      md = EVP_sha512();
+      break;
+    default:
+      throw std::runtime_error("SHADigest: unsupported size");
```
Comment (rgacogne) [inline pdns/sha.hh:78]: This leaks `mdctx`. Perhaps we could wrap it in a `unique_ptr` right away?

File: pdns/sha.hh [hunk 4]
```diff
@@ -51,3 +52,67 @@ inline std::string pdns_sha512sum(const std::string& input)
...
+
+  ~SHADigest()
+  {
+    // No free of md needed afaik
+    if (mdctx != nullptr) {
+      EVP_MD_CTX_free(mdctx);
+    }
+  }
+
+  void process(const std::string& msg, size_t sz)
```
Comment (rgacogne) [inline pdns/sha.hh:93]:
```suggestion
  void process(const std::string& msg)
```

File: pdns/sha.hh [hunk 5]
```diff
@@ -51,3 +52,67 @@ inline std::string pdns_sha512sum(const std::string& input)
...
+  std::string digest()
+  {
+    std::string md_value;
+    md_value.resize(EVP_MD_size(md));
+    unsigned int md_len;
+    if (EVP_DigestFinal_ex(mdctx, reinterpret_cast<unsigned char*>(md_value.data()), &md_len) == 0) {
+      throw std::runtime_error("SHADigest: finalize error");
+    }
+    if (md_len != md_value.size()) {
+      throw std::runtime_error("SHADigest: inconsisten size");
```
Comment (rgacogne) [inline pdns/sha.hh:109]:
```suggestion
      throw std::runtime_error("SHADigest: inconsistent size");
```

File: pdns/zonemd.cc [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+};
+
+typedef std::map<rrSetKey_t, rrVector_t, CanonrrSetKeyCompare> RRsetMap_t;
+
+void pdns::zonemdVerify(const DNSName& zone, ZoneParserTNG& zpt, bool& validationDone, bool& validationOK)
+{
+  validationDone = false;
+  validationOK = false;
+
+  // scheme,hasalgo -> duplicate,zonemdrecord
```
Comment (rgacogne) [inline pdns/zonemd.cc:33]:
```suggestion
  // scheme,hashalgo -> duplicate,zonemdrecord
```

File: pdns/pdnsutil.cc [hunk 7]
```diff
@@ -3154,7 +3200,7 @@ try
   }
   else if (cmds.at(0) == "unset-nsec3") {
     if(cmds.size() < 2) {
-      cerr<<"Syntax: pdnsutil unset-nsec3 ZONE"<<endl;
+      cerr<<"Syntax: pdnsutil unset-nsec3 ZON"<<endl;
```
[the remark this answers is not in the document]
Comment (omoerbeek) [reply pdns/pdnsutil.cc:3203]: Dunno what happened there

File: pdns/sha.hh (hunk 1 above)
Comment (omoerbeek) [reply pdns/sha.hh]: `Defines a type of object to be thrown as exception. It reports errors that are consequence of attempt to access elements out of defined range.`
So close, but not really it.

File: pdns/sha.hh [hunk 8]
```diff
@@ -51,3 +52,64 @@ inline std::string pdns_sha512sum(const std::string& input)
   SHA512(reinterpret_cast<const unsigned char*>(input.c_str()), input.length(), result);
   return std::string(result, result + sizeof result);
 }
+
+namespace pdns
+{
+class SHADigest
```
Comment (chbruyand) [inline pdns/sha.hh:58]: Maybe you could disable default constructor ?

**Review by chbruyand — APPROVED** [review]
Didn't dive too much into `zonemdVerify()` as I don't *know* ZONEMD and found it a bit hard to read (lots of abbreviations that are lowercased/uppercased), but LGTM as it's tested and the API is readable

**Comment by omoerbeek** [comment]
> Didn't dive too much into `zonemdVerify()` as I don't _know_ ZONEMD and found it a bit hard to read (lots of abbreviations that are lowercased/uppercased), but LGTM as it's tested and the API is clear and speaks for itself

Good point, I'll rename some vars

File: pdns/sha.hh (hunk 8 above)
Comment (omoerbeek) [reply pdns/sha.hh:58]: Default constructor is implicitly deleted if another one exists I think

File: pdns/sha.hh [hunk 9]
```diff
@@ -51,3 +52,64 @@ inline std::string pdns_sha512sum(const std::string& input)
...
+      throw std::invalid_argument("SHADigest: unsupported size");
+    }
+    if (EVP_DigestInit_ex(mdctx.get(), md, NULL) == 0) {
+      throw std::runtime_error("SHADigest: init error");
+    }
+  }
+
+  ~SHADigest()
+  {
+    // No free of md needed afaik
```
Comment (Habbie) [inline pdns/sha.hh:87]: I agree - so let's remove the afaik and mention the unique_ptr?

File: pdns/zonemd.cc [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+#include "zoneparser-tng.hh"
+
+typedef std::pair<DNSName, QType> RRSetKey_t;
+typedef std::vector<std::shared_ptr<DNSRecordContent>> RRVector_t;
+
+struct CanonRRSetKeyCompare : public std::binary_function<RRSetKey_t, RRSetKey_t, bool>
+{
+  bool operator()(const RRSetKey_t& a, const RRSetKey_t& b) const
+  {
+    // FIXME surely we can be smarter here
```
Comment (Habbie) [inline pdns/zonemd.cc:15]: I believe this is my FIXME, I'll have a go at it

File: pdns/zonemd.cc [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+
+  DNSResourceRecord dnsResourceRecord;
+
+  // Get all records and remember RRSets and TTLs
+  while (zpt.get(dnsResourceRecord)) {
+    if (!dnsResourceRecord.qname.isPartOf(zone) && dnsResourceRecord.qname != zone) {
+      continue;
+    }
+    if (dnsResourceRecord.qtype == QType::SOA && soaRecordContent) {
+      // XXX skip extra SOA?
```
Comment (Habbie) [inline pdns/zonemd.cc:54]: We should handle or remove this comment. I'm leaning towards remove.

File: pdns/zonemd.cc [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+      throw PDNSException(err);
+    }
+    if (dnsResourceRecord.qtype == QType::SOA && dnsResourceRecord.qname == zone) {
+      soaRecordContent = std::dynamic_pointer_cast<SOARecordContent>(drc);
+    }
+    if (dnsResourceRecord.qtype == QType::ZONEMD && dnsResourceRecord.qname == zone) {
+      auto zonemd = std::dynamic_pointer_cast<ZONEMDRecordContent>(drc);
+      auto inserted = zonemdRecords.insert({pair(zonemd->d_scheme, zonemd->d_hashalgo), {zonemd, false}});
+      if (!inserted.second) {
+        // Mark as duplicate;
```
Comment (Habbie) [inline pdns/zonemd.cc:76]: CodeQL thinks this is code; perhaps removing the ; will fix that :)

**Review by Habbie — APPROVED** [review]
Awesome work. I posted a few nits, but nothing major.

File: pdns/zonemd.cc (hunk 10 above)
Comment (Habbie) [reply pdns/zonemd.cc:15]: Ah yes, can't use `std::tie` because of the custom comparison function. I'll leave it for now..

**Pull request merged**